### ISSUE TITLE
uiTree control

### ISSUE THREAD

**Issue by cody271** [libui-ng/libui-ng#130]
[GtkTreeView](https://docs.gtk.org/gtk3/class.TreeView.html) on `unix`:
![image](https://user-images.githubusercontent.com/11763424/182323874-58caae70-8125-4a31-92a0-130af7c8f4b1.png)

[NSOutlineView](https://developer.apple.com/documentation/appkit/nsoutlineview) on `darwin`:
<img src="https://user-images.githubusercontent.com/11763424/182324718-625c35fe-fb2b-4ef0-ad96-bee1d33b79bb.png" alt="NSOutlineView" width="256" height="144">

[Tree-View](https://docs.microsoft.com/en-us/windows/win32/controls/tree-view-controls) on `windows`:
![image](https://user-images.githubusercontent.com/11763424/182325906-8f27e9b3-3bb0-4f75-862a-de0510d66eda.png)

**Comment by szanni** [comment]
> [Tree-View](https://docs.microsoft.com/en-us/windows/win32/controls/tree-view-controls) on `windows`

To also look into: table style tree views for windows, like in [SWT](https://www.eclipse.org/swt/):
![image](https://user-images.githubusercontent.com/286606/183260175-643819d0-6f9a-4613-9cfc-3a60e5085581.png)

While win32 does not seem to provide such functionality natively, it looks like implementing such a feature is possible (yes, a lot of hard work). I believe it might be worth while to have both:

- Something akin to the windows native TreeView for simple side panes
- A Gtk table style tree view implementation for more complex software

**Comment by jwells89** [comment]
For implementation of a table style tree view as mentioned by @szanni on macOS, there's `NSOutlineView` that provides this functionality natively.

**Comment by cody271** [comment]
> For implementation of a table style tree view as mentioned by @szanni on macOS, there's `NSOutlineView` which provides this functionality natively.

Yes `NSOutlineView` is the way, as [documented](https://github.com/libui-ng/libui-ng/issues/130#issue-1325459237) above. ;)

**Comment by cblc** [comment]
Are there any ideas on how the API should be for trees?

**Comment by cody271** [comment]
> Are there any ideas on how the API should be for trees?

No work has been done on `uiTree` yet, we've been focused on `uiTable` first.

**Comment by nptr** [comment]
I did experiment with creating a tree control. It is far from finished, hardly functional and just for Windows.
The idea behind the tree control is similar to that of the table control, where the user supplies a data model.

Because I likely will not continue working on it, I'll post the link to my work in hopes someone want to take this further:
[https://github.com/nptr/libui-ng/commits/treeview2/](https://github.com/nptr/libui-ng/commits/treeview2/)

Sorry for not providing the code in a nicer format (patch, pull request). Maybe some day.
